docs/design/clustering.md

```diff
@@ -0,0 +1,56 @@
+  3. **Caller AuthN/Z** The nodes publish events and so must be authenticated to the master. Ideally this authentication is specific to each node so that authorization can be narrowly scoped.  The details of the work to run (including things like environment variables) might be considered sensitive and should be locked down also.
+
+## Current Implementation
+
+A central authority (generally the master) is responsible for determining the set of machines which are members of the cluster.  Calls to create and remove worker nodes in the cluster are restricted to this single authority, and any other requests to add or remove worker nodes are rejected. (1.i).
+
+Communication from the master to nodes is currently over HTTP and is not secured or authenticated in any way.  (1.ii, 1.iii).
+
+The location of the master is communicated out of band to the nodes.  For GCE, this is done via Salt.  Other cluster instructions/scripts use other methods. (2.i)
+
+Currently most communication from the node to the master is over HTTP.  When it is done over HTTPS there is currently no verification of the cert of the master (2.ii).
+
+Currently, the node/kubelet is authenticated to the master via a token shared across all nodes.  This token is distributed out of band (using Salt for GCE) and is optional.  If it is not present then the kubelet is unable to publish events to the master. (2.iii)
+
+Our current mix of out of band communication doesn't meet all of our needs from a security point of view and is difficult to set up and configure.
+
+## Proposed Solution
+
+The proposed solution will provide a range of options for setting up and maintaining a secure Kubernetes cluster.  We want to both allow for centrally controlled systems (leveraging pre-existing trust and configuration systems) or more ad-hoc automagic systems that are incredibly easy to set up.
+
+The building blocks of an easier solution:
+
+* **Move to TLS** We will move to using TLS for all intra-cluster communication.  We will work to explicitly distributing and trusting the CAs that should be trusted for each link.  We will also use client certificates for all AuthN.
+* [optional] **API driven CA** Optionally, we will run a CA in the master that will mint certificates for the nodes/kubelets.  There will be pluggable policies that will automatically approve certificate requests here as appropriate.
+  * **CA approval policy**  This is a pluggable policy object that can automatically approve CA signing requests.  Stock policies will include `always-reject`, `queue` and `insecure-always-approve`.  With `queue` there would be an API for evaluating and accepting/rejecting requests.  Cloud providers could implement a policy here that verifies other out of band information and automatically approves/rejects based on other external factors.
+* **Scoped Kubelet Accounts** These accounts are per-minion and (optionally) give a minion permission to register itself.
```

Comment:
Practically speaking, the need to say "the kubelet is allowed to do things that are specific to its inherent identity" is similar to service accounts or other special users.  It would be ideal if we could somehow tie identity to policy in a way that a single policy represents what an identity of form X can do.  In the long run, we expect the kubelet policy, or extension plugins, to have well defined rights and behaviors.  It seems better to make that possible to do in an easily understandable way(in the long run).  Does the kubelet policy really need to be that flexible?  We might want to add to what it can do, but its core policy is likely a fundamental aspect of the system.

Possibly that calls for something like a special policy rule that applies to a kubelet identity with a default checker that can have overrides.

```
Policy:
   any-identity-that-matches-this-regex: kubelet@<host>
      acts-as-kubelet
      allow fooresource GET <...>
```